crypto/s2n_certificate.c

```diff
@@ -532,3 +532,53 @@ s2n_cert_private_key *s2n_cert_chain_and_key_get_private_key(struct s2n_cert_cha
     PTR_ENSURE_REF(chain_and_key);
     return chain_and_key->private_key;
 }
+
+int s2n_get_cert_chain_length(struct s2n_cert_chain_and_key *chain_and_key, uint32_t *cert_length)
+{
+    POSIX_ENSURE_REF(chain_and_key);
+
+    struct s2n_cert *head_cert = chain_and_key->cert_chain->head;
+    POSIX_ENSURE_REF(head_cert);
+    *cert_length = 1;
+    struct s2n_cert *next_cert = head_cert->next;
+    while (next_cert != NULL) {
+        *cert_length += 1;
+        next_cert = next_cert->next;
+    }
+
+    return S2N_SUCCESS;
+}
+
+int s2n_get_cert_from_cert_chain(struct s2n_cert_chain_and_key *chain_and_key, struct s2n_cert **out_cert,
+                                 uint32_t cert_idx)
+{
+    POSIX_ENSURE_REF(chain_and_key);
+    POSIX_ENSURE_REF(out_cert);
+
+    uint32_t cert_chain_length = 0;
+    POSIX_GUARD(s2n_get_cert_chain_length(chain_and_key, &cert_chain_length));
+    POSIX_ENSURE_LT(cert_idx, cert_chain_length);
```

Comment:
The reason `s2n_get_cert_chain_length` exists  as a separate function is because our customers would also use it to obtain the length of the cert chain.  Writing our own logic to obtain the length is duplicating this function's logic.

Reply:
I understand that it needs to exist as a separate function, but I don't think we benefit from reusing it here.

You need to iterate through the cert chain anyway to retrieve the right certificate. There's no benefit to iterating all the way through it first to get the length. You're not reducing code duplication; the loop still exists and you're still validating that next isn't null.